keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2140 @@
+}
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes. Besides the name, exactly one field must be set.
+// +k8s:deepcopy-gen=true
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name"`
+
+	// Basic defines one device instance.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Basic *BasicDevice
+
+	// Composite defines one composite device instance.
+	//
+	// +optional
+	// +oneOf=deviceType
+	Composite *CompositeDevice `json:"composite,omitempty"`
+}
+
+// CompositeDevice defines one device instance.
+type CompositeDevice struct {
+	// Includes defines the set of device mixins that this device includes.
+	//
+	// The propertes of each included mixin are applied to this device in
+	// order. Conflicting properties from multiple mixins are taken from the
+	// last mixin listed that contains them.
+	//
+	// The maximum number of mixins that can be included is 8.
+	//
+	// +optional
+	Includes []DeviceMixinRef `json:"includes,omitempty"`
+
+	// ConsumesCapacityFrom defines the set of devices where any capacity
+	// consumed by this device should be pulled from. This applies recursively.
+	// In cases where the device names itself as its source, the recursion is
+	// halted.
+	//
+	// Conflicting capacities from multiple devices are taken from the
+	// last device listed that contains them.
+	//
+	// The maximum number of devices that can be referenced is 8.
+	//
+	// +optional
+	ConsumesCapacityFrom []DeviceRef `json:"consumesCapacityFrom,omitempty"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// To ensure this uniqueness, attributes defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting attributes from those provided via mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Attributes map[QualifiedName]DeviceAttribute `json:"attributes,omitempty"`
+
+	// Capacity defines the set of capacities for this device.
+	// The name of each capacity must be unique in that set.
+	//
+	// To ensure this uniqueness, capacities defined by the vendor
+	// must be listed without the driver name as domain prefix in
+	// their name. All others must be listed with their domain prefix.
+	//
+	// Conflicting capacities from those provided via mixins are
+	// overwritten by the ones provided here.
+	//
+	// The maximum number of attributes and capacities combined is 32.
+	//
+	// +optional
+	Capacity map[QualifiedName]DeviceCapacity `json:"capacity,omitempty"`
+}
+
+// DeviceMixinRef defines a reference to a device mixin.
+type DeviceMixinRef struct {
```

Comment:
I'm happy t just ref by string. I only did it this way to make it extendable in case we saw a use for it in the future. If you think we should just flatten this to a string, I am fine with that.